chatmeleon-server/src/controllers/message-controller.ts

```diff
@@ -1,114 +1,252 @@
-  }
-
-  await prisma.message.create({
-    data: {
-      body,
-      image,
-      conversation: {
-        connect: {
-          id: conversationId,
-        },
-      },
-      sender: {
-        connect: {
-          id: senderId,
-        },
-      },
-    },
-  });
+  prisma
+    .$transaction(async (tx) => {
+      try {
+        // Validate conversation existence
+        await tx.conversation.findFirstOrThrow({
+          where: {
+            id: conversationId,
+            userIds: { has: senderId },
+          },
+        });
+
+        // Create the new message within the transaction
+        const newMessage = await tx.message.create({
+          data: {
+            body,
+            image,
+            conversation: {
+              connect: {
+                id: conversationId,
+              },
+            },
+            sender: {
+              connect: {
+                id: senderId,
+              },
+            },
+          },
+        });
+
+        // Update conversation with new message ID and timestamp
+        await tx.conversation.update({
+          where: {
+            id: conversationId,
+          },
+          data: {
+            lastMessageAt: newMessage.createdAt,
+            lastMessage: newMessage.body,
+            messagesIds: { push: newMessage.id },
+          },
+        });
+
+        // Send success response on successful transaction
+        return res.status(200).type("text/plain").send("OK");
+      } catch (error) {
+        // Rethrow the error to trigger the transaction rollback
+        throw error;
+      }
+    })
+    .catch((error) => {
+      // Log the error and send a 403 Unauthorized response
+      console.error("Transaction failed:", error);
+      res
+        .status(403)
+        .type("text/plain")
+        .send("Unauthorized: User is unauthorized");
+    });
+};
+
+// Controller for creating a new original conversation with initial messages
+const createOriginalConversationAndFirstMessages = async (
+  req: Request,
+  res: Response
+) => {
+  // Extract data from request body and authentication
+  const { relatedUserId, body, image } = req.body;
+  const { userId: senderId } = req.auth!;
+
+  // Utilize a Prisma transaction to ensure atomicity and data consistency
+  prisma
+    .$transaction(async (tx) => {
+      try {
+        // Check for existing original conversation between the two users
+        const existingConversation = await tx.conversation.findFirst({
+          where: {
+            userIds: {
+              hasEvery: [senderId, relatedUserId], // Must include both sender and recipient
+            },
+            isOriginal: true, // Only consider original conversations
+          },
+        });
+
+        if (existingConversation) {
+          // Conversation already exists, return an error code
+          return res
+            .status(409)
+            .type("text/plain")
+            .send("Conflicted: Original conversation already existed");
```

Comment:
Move http response logic outside the transaction. The transaction can throw a error or something at this point to return control